File: Services/pkg/domain/user_role.go
package domain

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	helpers "github.com/zercle/gofiber-helpers"
)

type UserRoleRepository interface {
	Migrate() error
	CreateUserRole(ctx context.Context, userRole models.UserRole) *helpers.ResponseError
	GetUserRole(ctx context.Context, id uint) (*models.UserRole, *helpers.ResponseError)
	GetUserRoles(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.UserRole, *models.Pagination, *models.Search, *helpers.ResponseError)
	GetUserRolesByUserID(ctx context.Context, userID uint) ([]models.UserRole, *helpers.ResponseError)
	UpdateUserRole(ctx context.Context, id uint, userRole models.UserRole) *helpers.ResponseError
	DeleteUserRole(ctx context.Context, id uint) *helpers.ResponseError
	DeleteUserRolesByUserID(ctx context.Context, userID uint) *helpers.ResponseError
}

type UserRoleService interface {
	CreateUserRole(ctx context.Context, userRole models.UserRole) []helpers.ResponseError
	GetUserRole(ctx context.Context, id uint) (*models.UserRole, []helpers.ResponseError)
	GetUserRoles(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.UserRole, *models.Pagination, *models.Search, []helpers.ResponseError)
	GetUserRolesByUserID(ctx context.Context, userID uint) ([]models.UserRole, []helpers.ResponseError)
	UpdateUserRole(ctx context.Context, id uint, userRole models.UserRole) []helpers.ResponseError
	DeleteUserRole(ctx context.Context, id uint) []helpers.ResponseError
	DeleteUserRolesByUserID(ctx context.Context, userID uint) []helpers.ResponseError
}

File: hono-clean-architecture/pkg/domain/user_role.ts
import { UserRole, Pagination, Search } from "../models";
import { ResponseError } from "../../internal/infrastructure/custom_error";

export interface UserRoleRepository {
  migrate(): Promise<void>;
  createUserRole(userRole: { userId: number; roleId: number }): Promise<ResponseError | null>;
  getUserRole(id: number): Promise<{ userRole: UserRole | null; error: ResponseError | null }>;
  getUserRoles(
    pagination: Pagination,
    search: Search
  ): Promise<{
    userRoles: UserRole[];
    pagination: Pagination;
    search: Search;
    error: ResponseError | null;
  }>;
  deleteUserRole(id: number): Promise<ResponseError | null>;
  getUserRolesByUserId(
    userId: number
  ): Promise<{ userRoles: UserRole[]; error: ResponseError | null }>;
}

export interface UserRoleService {
  createUserRole(userRole: { userId: number; roleId: number }): Promise<ResponseError[]>;
  getUserRole(id: number): Promise<{ userRole: UserRole | null; errors: ResponseError[] }>;
  getUserRoles(
    pagination: Pagination,
    search: Search
  ): Promise<{
    userRoles: UserRole[];
    pagination: Pagination;
    search: Search;
    errors: ResponseError[];
  }>;
  deleteUserRole(id: number): Promise<ResponseError[]>;
}
